### ISSUE TITLE
Yarn over npm?

### ISSUE THREAD

**Issue by shouston3** [dwyl/technology-stack#39]
We are currently using just npm to manage our dependencies, but I think there are better alternatives such as [yarn](https://yarnpkg.com/).

I have seen that we're not using it in any of our projects. Why is this?

[I can't see any reason to not be using it](https://medium.com/@nikjohn/facebooks-yarn-vs-npm-is-yarn-really-better-1890b3ea6515#.29e0pdg1e)

**Comment by nelsonic** [comment]
@shouston3 great question! (_thanks for opening this discussion thread_!) 😍 

***Yes***, Yarn is on our radar: https://github.com/nelsonic/nelsonic.github.io/issues/234 🤔 
_we_ haven't seen the ***need*** for it, yet. 
have ***you*** felt the ***need*** for it on one of your projects?
e.g: do you have a project where installing dependencies takes longer than **20 Seconds**?
is it _affecting_ your _productivity_ e.g: by slowing down the "CI" build for a PR which means you cannot deliver your work on schedule...?
These would both be _legitimate_ reasons to consider using an "_alternative_" to NPM, 
using Yarn just because it's _new_ and has a nice website with ***3 Key Advantages*** (`#Marketing101`) is _not_ (_usually_) a good reason to adopt something. if you're feeling ***pain*** using NPM that's different... 😢 

> I _acknowledge_ that Yarn is "_faster_" than NPM for downloading packages.
It employs a _far_ more **effective _caching_ mechanism**. **no question** there.
I don't tend to _re-install_ any dependencies in my project(s) very often unless there is a _reason_ to upgrade something and then I would need to install a ***new*** version in _any_ case, whereby Yarn offers _zero_ advantage over NPM.

Unless you are working on _many_ projects that have similar dependencies 
e.g. you're building several React components in separate repositories, publishing them individually to NPM and then including them back into a "_main_" project ... 
that would be a _good_ use-case for Yarn. 👍 
_Most_ people @dwyl are working on _one_ project at a time. 
_Occasionally_ we "_overlap_" from one project to the next but _rarely_ because **context switching** is **cognitively _expensive_**: http://geekswithblogs.net/MarkPearl/archive/2012/07/17/the-human-multitasking-problem-ndash-the-cost-of-a-context.aspx

## Facebook are _Creating_ an _Artificial_ "Need" for a New Package Manager by using _Bloated_ Dependencies instead of being Efficient ...

> I _can_ see the "_need_" for Yarn in the context of **installing _many_ large dependencies**. _however_ ...

@dwyl We don't routinely use ***mega-depenencies*** like "Babel" which we _agree_ takes _ages_ to install!! 
![image](https://cloud.githubusercontent.com/assets/194400/22400099/1e6bac2a-e5a4-11e6-8cdc-a1374a667085.png)

> So... I (_personally_) don't feel the _need_ to install _another_ global package on my system ...

Facebook are "solving" a "problem" **they _created_** by building _huge_ modules in the _first place_ and **_fragmenting_ the node.js community** instead of (_patiently_) **contributing _back_** to NPM. 

***Thankfully, None*** of the projects/products @dwyl suffer from the "_takes ages to install_" problem.
Because we build _lightweight_ things that have **as _few_ `dependencies` as _possible_** for both ***sanity and security*** reasons ... http://www.theregister.co.uk/2016/03/23/npm_left_pad_chaos/ ... 😧 

What "FB Engineering" ***could*** (***should***) have done is take their _Vast_ Capital and ***approach*** the _lovely_ (_talented and super dedicated_!) people at **NPM Inc** (_@isaacs and @seldo are super approachable and community-minded!_) with a ***proposal*** ... _for example_:

> "_Hey NPM, we know you are over-worked trying to keep the most popular package manager on the Internet running and haven't got that much time for improvements ... would you consider letting us pay you guys a million bucks if you let us partner with you on improving NPM using all our node.js resources...?_"

## _Preliminary_ Conclusion

We don't ***need*** Yarn, yet. 
Please feel free to _monitor_ how long it takes to install your dependencies for a given DWYL project and if it goes above **15 seconds** (_or how ever long it takes you to read the next user story in the backlog for the project_), leave a new comment on this thread so we can _re-consider_ Yarn! 👍 

Thanks!! ❤️  ✅ 

<br /> <br /> <br />

# tl;dr (_warning: contains history lesson_)

Yarn is _technically_ "_better_" than NPM for the reasons described on the _slick_ website...
but it _neglects_ to mention that the _reason_ NPM is over-worked is ***because*** Babel, Webpack, React etc. are **_bloated_** packages which ***cause*** huge **strain** on the **NPM infrastructure** which is "***free***" (_and always has been_) to use for the _community_!! 
has ***any*** "node.js deveveloper" you know ever ***paid*** NPM a cent to _use_ their service...?meanwhile they have hundreds of thousands of dollars in bandwidth and storage costs... 
but _instead_ of helping to _alleviate_ the problem e.g. by contributing more resources to NPM Facebook decided to re-write it and _crucially_ ***re-brand*** it... that's the _key_ part! ***re-branding***...

Facebook (_and handful of other "_Big_" companies_) always want to ***own*** the system rather than _use_ (_and help improve_) the already pretty decent system that exists. this is their way of trying to "brand" something that they did **not invent** and were **not** a part of creating but have decided to "_appropriate_" now because:
![image](https://cloud.githubusercontent.com/assets/194400/22400019/3dbabbae-e5a2-11e6-96c9-d14f5864cd74.png)

> **note**: I did not _create_ that "**meme**", it's _pervasive_ on the net.
just read: https://www.theguardian.com/technology/2016/dec/16/facebook-copies-snapchat-feature-for-15th-time imagine you're the developer working (_really_) hard to release a new thing/feature and Facebook just swooped in and shamelessly copied it often taking the "_credit_" for inventing it, how would you `feel`...?
***everyone*** who has been in Web development knows that Fb have _systematically_ "_claimed_" every good idea that any _other_ "social network" company has created. That's just lame.

> There _must_ be a _reason_ "Zuck" was portrayed the way he was in "The Social Network", 
it's not just "Hollywood Drama" to "_sensationalise_" the events ...  
https://youtu.be/VKnNUYknsuQ and https://youtu.be/xdiFzcpmmJc

> FB have a _culture_ of "_improving_" on other people's ideas, gaining a competitive advantage (_because they have more capital to spend on customer acquisition_) and then "_fighting it out in court_" which is _always_ more difficult for the "_little guy_"... It worked for Microsoft against Netscape: https://youtu.be/m_2m1qdqieE (_if you have never watched it, add it to your "watch later" list cause it's a "**master class**" in **legal gymnastics**, Bill Gates is **unquestionably** a **genius** and having studied "pre law" played a **big** role in the success of his software company!_)
... and "Zuck" is _every bit_ as "_savvy_" as Bill Gates was when it comes to using the _best_ legal team to achieve his personal ambitions ... 

> There _must_ be a _reason_ why "Zuck" has a _team_ of PR people "_managing_" his "_public image_": https://www.bloomberg.com/news/articles/2017-01-18/this-team-runs-mark-zuckerberg-s-facebook-page taking and carefully releasing the soft-focus photo op "official" photos ... 

> This _might_ not be "_relevant_" to the discussion about Yarn _specifically_, but it is relevant to the _wider_ discussion about using "Developer Tools/Libraries/Frameworks" _made_ by Facebook: https://github.com/dwyl/learn-react/issues/23 beyond discussing the ***technical merit*** of the tool we ***have*** to consider the ***context*** of it ... I don't have a "_technical objection_" to using FB tools, in fact I ***like*** GraphQL: https://github.com/dwyl/learn-graphQL 

Let's _evaluate_ Yarn on it's ***own*** merits. and determine if we have a ***need*** to use it. 👍

**Comment by iteles** [comment]
Thanks for raising this @shouston3 👍 :tada: A very legitimate question and I'm glad you've allowed us to have a place to address it! 
If you do think there is a *need* to move to Yarn, please don't hesitate to let us know as @nelsonic says :blush:

**Comment by shouston3** [comment]
@nelsonic thank you so much for the insight on how yarn probably came about, and from my subsequent reading around the topic, you are right that initially yarn came about from high volumes of dependencies and slow build times, but also nondeterministic installs (which myself and others have felt the unnecessary pain of) and [security issues](https://techcrunch.com/2016/10/11/facebook-partners-with-google-others-to-launch-a-new-javascript-package-manager/).

Although it has been interesting reading up on the history of the npm and facebook teams and how this tool came to be, I would prefer to focus on the pros and cons of using this tool.

### Need?

I don't know why there is such a focus on the "need" for it, it is simply an alternative tool to use over the npm cli to improve your workflow.

I think a better question to ask would be, do the benefits of using yarn outweigh the costs?

With yarn, there is also practically no learning curve, you can be fully up and running in a few seconds (with a couple of seconds install time (`brew install yarn`) and a couple more seconds [learning the interface](https://yarnpkg.com/en/docs/migrating-from-npm#toc-cli-commands-comparison) and your good to go)

Also, any project that implements yarn won't force anyone to have yarn installed globally or to use yarn at all, they can still `npm install` and everything will work fine 👍 .

(But you still might as well run `brew install yarn && yarn` instead, as it would install faster and not fill up your terminal with stuff you will never read, etc).

To answer your specific questions:

#### Have you felt the need for it on one of your projects?
Not the "need", I can manage without it, but over the last few months I have been using it when I can because it speeds up my workflow with (as of yet) no downsides.

#### Do you have a project where installing the dependencies takes longer than 20 seconds?
For Dwyl projects, large install times are rare, but my last project took more than a minute to install and [some other dwyl projects have similar install times](https://travis-ci.org/TheScienceMuseum/collectionsonline)

(Also perhaps this is caused by the specs on my macbook air, but running `npm install` in most projects takes longer than 20 seconds to run).

#### Is it slowing down your productivity?
Yes

### However

Yarn is really new and your caution is definitely warranted (and your not alone [here](https://github.com/nodejs/node/issues/9161) and [here](https://github.com/nodejs/docker-node/issues/243)).

Although it is actively maintained, there is quite a lot of [open issues](https://github.com/yarnpkg/yarn/issues).

And perhaps it is overkill for small projects

### Conclusion

Could we try it out in one or two projects and then evaluate if the benefits outweigh the costs?

**Comment by nelsonic** [comment]
@shouston3, good question. _Why_ should we focus on a "***need***"...? 🤔 

> _Time is **finite**. given the **choice** is between adding a **user-requested feature** vs. a dev `technical task` with **limited user-benefit**, the **user story** gets **priority every time**_. 

when making a change to an _existing_ project we **always _focus_** on the **_User_ need**, because otherwise we would end up with a project that gets _re-written several times_ (_at high expense to the client_) without **any _direct_ benefit** to the end-user... i.e: ["FOCH"](https://david-dm.org/foundersandcoders/sail-back) ...

so, we identify a ***need*** first and work from there. 
Not: "_oh, there's this [**shiny new thing**](http://tedrubin.com/beware-of-the-shiny-new-toy-syndrome-people-are-social-not-platforms) that has a slick website and **so many** "advantages" over the "old" (boring) tool, so, we should **use it everywhere**!!_" 😕 

> _it sounds like you **might** have a **pain point** on a slower-to-install project..._
> _**document the pain** with **specific metrics** and_ https://en.wikipedia.org/wiki/Root_cause_analysis

I _understand_ the _desire_ (_or curiosity_) to try a new tool 
and also _acknowledge_ that Facebook added Checksums for "_security_", 
but remember that you are now ***trusting*** Facebook to control your dependencies ... 
the same company that sells your eyeballs & personal data to the highest bidder ... 

> "***security***" is a _compelling_ "reason" FB are giving for using Yarn.
They know they have to give a "USP" to get devs to hand over all their download/usage stats for NPM packages ... this is data FB ***will*** use to **recruit** the people who they discover as being the biggest users of their "stack"... it's a _brilliant_ recruitment strategy. 💡 
and _anyone_ who does not see that is _naive_ about FB's true intentions.
ask @eliascodes or @bradreeder how much they are _itching_ to install & use Yarn _everywhere_!

Read: https://www.facebook.com/help/152637448140583 do you _honestly believe_ this...?!
If Facebook are _not_ selling your personal data, why would ***Harvard University Press*** publish this book: https://www.amazon.com/gp/product/0674504577
Summarised in: http://uk.businessinsider.com/facebook-google-information-nsa-iphone-android-data-personal-2016-2

> Do you think for a _nanosecond_ that Yarn is not sending all the _usage_ stats for _exactly_ what you are downloading and what you are working on to FB HQ...?

***Yes***, the science museum is using Babel... https://travis-ci.org/TheScienceMuseum/collectionsonline#L503 and it makes me sad. 😢 
but I didn't want to "_argue_" ... some people feel that ES6 adds "_value_" to their work...
If you are asking me: would I do that ***again***? the answer is a **_resounding_: No**! 
I would use https://github.com/dwyl/goodparts all-day-long and **reduce** the number of _Huge_ **dependencies** to an ***absolute minimum***. 

> If you want to `try` and use Yarn for The Science Museum CI 
and see if you can reduce the Build time from ***11 mins*** I would be ***delighted***!! 🎉  
(_on success this would be 100% welcome and I'd be stoked to pay an invoice for this work!_)

## Cost-Benefit Analysis

**Yes**, you will get a increase in speed installing dependencies using Yarn.

> ***Please Quantify*** the loss in productivity in hours per day/week/month/year. 
> and _specifically_ note down all the instances where you have been _unable_ to get _any_ (_other_) work done because NPM has _failed_ you. 😞 

I install things ***maybe*** **once every couple of weeks** when there is an update to a dependency in a project I'm working. `npm install` (_and do something `else` like reading a PR/issue `while` it's doing it's thing, so I'm **never waiting**... if people are `that` bad at saving time-bound tasks that require no user-input for when they have a bathroom/water break, or are `that` **impatient** that they cannot wait for NPM to install something then I have to **wonder** what other corners are being cut in the name of "saving time"..._) `#PleaseExcuseTheRantItsNotDirectedAtYou` ...

> NPM works _flawlessly_ for me. Long Live NPM! ❤️  

By contrast I can see a ***glaring*** "_downside_" of installing a _Global_ Tool Made by an organisation **I _don't trust_**: 
![zuck-yeah-so-if-you-ever-need-info-about-anyone-2939913](https://cloud.githubusercontent.com/assets/194400/22406857/dd6b7728-e652-11e6-921b-70ec29556bb8.png)

`while` on the _surface_ the tool _sounds_ "_more secure_" because it has checksums...
adding a "proxy" (_man-in-the-middle for **all** my dependencies_) violates one of the most _basic_ laws of Computer Security: https://en.wikipedia.org/wiki/Trust_no_one_(Internet_security)

> Also calling something "_deterministic_" sounds really _sophisticated_, but it's BS.
I have been using NPM for 6 years and can count the number of times the Dependencies failed to install on ***one hand***. If your experience is _different_ **please _share_** which project you are feeling the pain on. (_thanks!_) 👍 

_Nothing_ against _other_ people using Yarn. 
Each person has a different level of comfort/trust with Facebook. 
_Personally_ I don't trust them. 
and giving them Global access to my laptop without a ***very good reason*** (_i.e a **need**_) 
is a **definite "_no no_"**... 

_Thankfully_, this isn't a "_problem_" I'm going to have for _much_ longer... ⏳ 
I'm absolutely _delighted_ with Elixir! https://github.com/dwyl/learn-elixir 😍 
and so **_incredibly_ excited** to be using Elm https://github.com/dwyl/learn-elm 
I _highly doubt_ Facebook will be meddling in Elixir/Elm at least for the _foreseeable future_ given their investment in React, Immutable.js, Flow, Babel, Redux (_Elm's poor cousin_), etc. 🙃 

Personally, I will `continue`will using ("_boring old_") NPM on my localhost even if it's "slower" ...
until I **_completely_ transition** to `elixir`. 

But _everyone_ `else` is welcome to `try` Yarn 
and send their node package download/usage data to Facebook
and act _surprised_ when the FB "Technical Recruiter" emails them... 👍 

It _boggles_ my mind how people take NPM for granted. 😢 
Sure Yarn looks/sounds "_better_" on the _surface_...
but most people _really_ aren't thinking through the "_true cost_" of routing all their dependency requests via https://registry.yarnpkg.com i.e. telling Facebook ***even more*** about you.

And `while` anyone can _counter-argue_ that Facebook "doesn't care" or they _can_ just read your dependency list from your `package.json` on GitHub to "_mine_" who is using their Stack ... 
I _urge_ everyone to `try` and think about this from the _sceptical_ perspective that _maybe_ they are collecting more data on all developers in the community so that they can _recruit_ people to work on their **distraction machine**. see: http://www.timewellspent.io/
and actually this is a _brilliant_ way of collecting mountains of metadata on all developers in the Node.js community ... I'm _really_ not going to fall into the trap of _endorsing_ Yarn without a _very_ good reason/ "***need***". 👍 

Love you Sam! thanks again for opening this discussion! ❤️

**Comment by shouston3** [comment]
That is not really a perspective I had thought about too much and as it seems to be [a popular one](https://github.com/dwyl/learn-react/issues/23), it makes sense to not implement yarn.

My issues lie with previous projects and can't really see these arising with the new stack.

I am going to continue to use yarn for personal projects as I literally have nothing I care about on my computer (should I still watch out?) and will happily ignore fb recruiters should they come knocking.

**Comment by Charlotteis** [comment]
I don't see how npm is a better choice over yarn in most instances.
It is broken in a _lot_ of ways. Yarns power is fast install of large amounts of packages, are you going to be installed 100s of packages a day repeatedly?

I don't like the idea of trying something new just because. npm is widely used and thus the barrier to entry on an informational stand point, is lower.

**Comment by shouston3** [comment]
Great points @Charlotteis, I think I'll put this issue out of it's misery

**Comment by isaacs** [comment]
@Charlotteis 

> I don't see how npm is a better choice over yarn in most instances.

Is that a typo?  It seems like this sentence is incongruent with the rest of what you're saying in your post.

In any event, thanks for the generous props and/or candid criticism, as appropriate ;)

**Comment by Charlotteis** [comment]
@isaacs typo indeed! hehe!

**Comment by Neats29** [comment]
@nelsonic I got here from the discussion on the things-of-interest channel. We might also be switching to Yarn at work, and I've been reluctant as I've never had any issues with modules taking too long to install. So I appreciate the points raised here.

You mentioned:

> ... and giving them Global access to my laptop...

Could you please point me to any links that may explain this in more detail?

Thank you.

**Comment by nelsonic** [comment]
@Neats29 if you don't have a _rule_ for who trust 
`.then` the choice of where you install executables from is a "_personal_ decision" 
(_i.e. up to each individual/team..._)

Have a look at the installation instructions for Yarn: https://yarnpkg.com/en/docs/install#mac-tab

![yarn installation](https://cloud.githubusercontent.com/assets/194400/23558656/4afcf774-002c-11e7-8018-f7a839f64479.png)

notice the `sudo`?

Look at how they install it on Ubuntu/Debian: 
https://github.com/yarnpkg/yarn/blob/d2865a91bc58c89a5e9e4e52a3b210e41e655add/scripts/bootstrap-env-ubuntu.sh#L6

Why `sudo`...? Why does it need "root" / "admin" access to my machine...?

Yarn is _probably_ perfectly "_safe_" for _now_...
but they are violating a _basic_ security principal: 
https://github.com/dwyl/learn-security#principal-of-least-priviledge

A library that _forces_ me to install with `sudo` is an _immediate_ "No".
Even if they _claim_ to be "_more secure_", 
which as far as I can tell from reading the entire source code of the package when it was first announced, all they are doing is re-writing urls from HTTP to _HTTPS_ and doing basic checksums which NPM already has. 

Again, each person should make up their own mind if they want to use Yarn.

# tl;dr

Sure, package caching is a _useful_ feature, but the 3 "USPs" of Yarn could/should have been PR'd into NPM but instead Yarn (_Facebook_) have merely re-packaged (_excuse the pun_) the NPM's hard work and _re-branded_ it so they can "_own_" after doing _none_ of the _difficult_ work of actually running a package repository!

They should have contributed the improvements up-stream instead of fragmenting the ecosystem.

Personally I'm _way_ more interested in using "***PETE***" for _all_ our future projects: 
https://github.com/dwyl/technology-stack/#the-pete-stack
Our use of node.js is tapering off _fast_ ... elixir is ❤️ 

I _was_ really excited about node.js in 2011 
because the idea of having "One Language" everywhere was _compelling_ for beginner friendliness. 
But as time has passed (_and dozens of new but [pointless "features"](https://github.com/dwyl/hq/issues/167#issuecomment-273228006) have been added_) it's lost it's shine. 
JS is _not_ the most "_elegant_", "_performant_" or "_self-documenting_" language.
We use JS because we _have_ to, not because it's "the best".
People are trying to make it "_better_" ...
but it's really a case of [sprinkling glitter](https://www.google.co.uk/search?q=sprinkle+brigade) ...
JS it's becoming more like Java with each release (_and that's not a good thing!_) there are **eight** ways of writing a function now. Eight! Why?! 😧 

Anyway... we all _know_ it's possible to write simpler code. 
The feedback from @dwyl people who are learning Elixir/Phoenix is _really_ positive:

![image](https://cloud.githubusercontent.com/assets/194400/23576509/6445e5ba-009f-11e7-8278-26ba87d55d89.png)

Instead of debating wether to use NPM or Yarn, stick with NPM for Node projects 
and learn "***PETE***" in your "spare time" 
so you can web Apps/APIs that handle **4x** the number of connections/requests 
compared Node.js/Express for the equivalent hardware.
Elixir/Phoenix/Elm enable us to 
**Write _considerably_ less code** that is _much_ easier to test, maintain & extend! 😍 

_Don't Settle. Build Better_! 🚀

**Comment by Neats29** [comment]
@nelsonic thank you for the detailed response.
